### PR TITLE
Add learning rate schedulers

### PR DESCRIPTION
**Pull request by gordicaleksa**
Refactored the learning rate schedulers code - we'll keep all definitions inside "schedulers.h" as per our offline discussion.

Supported LR schedulers:
* Cosine with warmup
* Cyclic triangular (https://arxiv.org/abs/1506.01186)
* Constant
* Linear with warmup

### CODE REVIEW

**Comment by karpathy** [comment]
So I think we'll want to identify the scheduler using a string e.g. "cosine" from argparse, and inside the schedulers file should `strcmp` and build the right one.

**Comment by karpathy** [comment]
Note there's also a previous PR on learning rate schedulers, including one that is trapezoidal.

**Comment by gordicaleksa** [comment]
Argh apologies i saw your todo list and just picked it up assuming no one has done it, plus it's such an orthogonal contribution i thought it would have been merged by now - will check the PRs before submitting the next time :)

**Comment by gordicaleksa** [comment]
https://github.com/karpathy/llm.c/pull/508 <- is this the one?

**Comment by eliebak** [comment]
Yes i think it was mine, but your PR is much nicer nw aha. It can be nice to identify the schedule from argparse, and if we use trapezoidal, you also have to define the number of decay steps that you want. I can close mine and we can add this features to this one?

**Comment by gordicaleksa** [comment]
Cool!

I can quickly implement the identification via a string. And happy to add trapezoidal as well. Let's see what Andrej thinks about the next steps.

**Comment by gordicaleksa** [comment]
Pushed a very ugly switching between the schedulers based on the input argument.

In order to make this closer to what we're used to in Python/C++ we basically have to implement a mini virtual function pointer logic in order to allow for object oriented features.

Happy to do it, and won't be hard, just wanted to double check the decision makes sense because we'll be dealing with func pointers and i don't think we've done previously in the codebase.

cc: @karpathy

**Comment by gordicaleksa** [comment]
It would look something along the lines of:
```
typedef struct LRScheduler {
    void (*initialize)(struct LRScheduler* self, double initial_rate);
    void (*step)(struct LRScheduler* self);
    double (*get_rate)(struct LRScheduler* self);
    double current_rate;
} LRScheduler;
```

File: train_gpt2.cu [hunk 1]
```diff
@@ -1546,6 +1552,28 @@ int main(int argc, char *argv[]) {
     Tokenizer tokenizer;
     tokenizer_init(&tokenizer, "gpt2_tokenizer.bin");
 
+    // set up learning rate scheduler
```
Comment (karpathy) [inline train_gpt2.cu:1555]: we could move this new block inside the schedulers file. and the block below as well.

Comment (gordicaleksa) [reply train_gpt2.cu:1555]: Done, simplified the code, it's ready.

File: llmc/schedulers.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
+/*
+Implements various learning rate schedulers.
+*/
+#ifndef SCHEDULERS_H
+
+#define SCHEDULERS_H
+
+#include <assert.h>
+#include <math.h>
+#include <cstring>
```
Comment (ngc92) [inline llmc/schedulers.h:10]: string.h, if you stick with C headers

File: llmc/schedulers.h [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    // warmup_iterations <- not used.
+    // train_num_batches <- not used.
+    // final_learning_rate_frac <- not used.
+    return scheduler->learning_rate;
+}
+
+// switch to the appropriate learning rate scheduler
+float get_learning_rate(LRSchedulerType lr_scheduler_type, LearningRateScheduler *scheduler, int step) {
+    float step_learning_rate;
+    if (lr_scheduler_type == LR_SCHEDULER_COSINE) {
```
Comment (ngc92) [inline llmc/schedulers.h:121]: switch/case?

Comment (gordicaleksa) [reply llmc/schedulers.h:121]: i don't think there is any advantage using it?

File: llmc/schedulers.h (hunk 2 above)
Comment (gordicaleksa) [reply llmc/schedulers.h]: fixed

File: llmc/schedulers.h (hunk 3 above)
Comment (ngc92) [reply llmc/schedulers.h:121]: with switch, you usually get IDE warnings if you forget one enum value, not sure if you get the same with ifs. Otherwise, its pretty much equivalent.

**Comment by karpathy** [comment]
forked this work into new PR
https://github.com/karpathy/llm.c/pull/627
made simplifications
added wsd
closing this one
